{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "66b9735b-6f09-4730-b692-b93feedd6ac4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e1f60ef5-8121-4cda-a12e-5ee17cf48e22",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>raw_name</th>\n",
       "      <th>xeno_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene25011|Xelaev18004747m</th>\n",
       "      <td>gene25011|Xelaev18004747m</td>\n",
       "      <td>Xelaev18004747m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene21250|Xetrov90028798m.L</th>\n",
       "      <td>gene21250|Xetrov90028798m.L</td>\n",
       "      <td>Xetrov90028798m.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene27977|Xelaev18004749m</th>\n",
       "      <td>gene27977|Xelaev18004749m</td>\n",
       "      <td>Xelaev18004749m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene26149|Xelaev18004750m</th>\n",
       "      <td>gene26149|Xelaev18004750m</td>\n",
       "      <td>Xelaev18004750m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene25611|Xelaev18004751m</th>\n",
       "      <td>gene25611|Xelaev18004751m</td>\n",
       "      <td>Xelaev18004751m</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                raw_name        xeno_symbol\n",
       "gene25011|Xelaev18004747m      gene25011|Xelaev18004747m    Xelaev18004747m\n",
       "gene21250|Xetrov90028798m.L  gene21250|Xetrov90028798m.L  Xetrov90028798m.L\n",
       "gene27977|Xelaev18004749m      gene27977|Xelaev18004749m    Xelaev18004749m\n",
       "gene26149|Xelaev18004750m      gene26149|Xelaev18004750m    Xelaev18004750m\n",
       "gene25611|Xelaev18004751m      gene25611|Xelaev18004751m    Xelaev18004751m"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Load the ANNData with symbols\n",
    "adata_xeno = sc.read_h5ad(\"../adata_xeno_with_symbols.h5ad\")\n",
    "adata_xeno.var.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3b2d6847-d089-4a9f-8db3-9dfa7a3fab7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(33985, 41560)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Xenopus AnnData genes\n",
    "adata_xeno.n_obs, adata_xeno.n_vars"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bd35aae4-57ea-45ed-85d3-a2d51c615602",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>xeno_symbol</th>\n",
       "      <th>human_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>trnt1</td>\n",
       "      <td>tRNA nucleotidyl transferase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>foxh1.2</td>\n",
       "      <td>forkhead box H1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nr5a2</td>\n",
       "      <td>nuclear receptor subfamily 5 group A member 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tbx1</td>\n",
       "      <td>T-box 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nr1d1</td>\n",
       "      <td>nuclear receptor subfamily 1 group D member 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  xeno_symbol                                   human_symbol\n",
       "0       trnt1                   tRNA nucleotidyl transferase\n",
       "1     foxh1.2                                forkhead box H1\n",
       "2       nr5a2  nuclear receptor subfamily 5 group A member 2\n",
       "3        tbx1                                        T-box 1\n",
       "4       nr1d1  nuclear receptor subfamily 1 group D member 1"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Load and parse the Xenbase ortholog file\n",
    "orth = pd.read_csv(\"../data/XenbaseGeneHumanOrthologMapping.txt\",\n",
    "                   sep=\"\\t\", header=None)\n",
    "\n",
    "orth = orth.rename(columns={0: \"xeno_entrez\",\n",
    "                            1: \"xenbase_id\",\n",
    "                            2: \"xeno_symbol_raw\",\n",
    "                            3: \"human_desc\"})\n",
    "\n",
    "orth[\"xeno_symbol\"] = orth[\"xeno_symbol_raw\"].astype(str)\n",
    "orth[\"human_symbol\"] = orth[\"human_desc\"].astype(str).str.split(\",\").str[0].str.strip()\n",
    "\n",
    "orth = orth.dropna(subset=[\"human_symbol\"])\n",
    "orth = orth[orth[\"human_symbol\"] != \"\"]\n",
    "orth[[\"xeno_symbol\", \"human_symbol\"]].head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6f8ef062-bd91-432a-a8db-b9014b663afb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15897, 6)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Xenbase ortholog rows\n",
    "orth.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d956b89a-3dbf-4f1e-8d7d-7c1358a62acf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(41188, 15884, 101)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Set of Xenopus gene symbols present in the scRNA-seq object\n",
    "xeno_symbols_data = set(adata_xeno.var[\"xeno_symbol\"])\n",
    "# Set of Xenopus gene symbols that have a human ortholog in the Xenbase table \n",
    "xeno_symbols_xenbase = set(orth[\"xeno_symbol\"])\n",
    "\n",
    "# Reports the no. of unique symbols in the scRNA data, Xenbase table, and shared between the two\n",
    "len(xeno_symbols_data), len(xeno_symbols_xenbase), len(xeno_symbols_data & xeno_symbols_xenbase)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "cd02a45a-ed9f-4c19-bd05-9735fc4927e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(101, 6)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Keeping only the Xenbase enteries whose Xenopus symbol actually occurs in our scRNA-seq data\n",
    "xeno_orth_sub = orth[orth[\"xeno_symbol\"].isin(xeno_symbols_data)].copy()\n",
    "# Check the remaining ortholog pairs after filtering\n",
    "xeno_orth_sub.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2af7c507-7161-42bb-ba67-d32b87b30fce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>xeno_symbol</th>\n",
       "      <th>human_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>csnk1a1</td>\n",
       "      <td>casein kinase 1 alpha 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>fgfr3</td>\n",
       "      <td>fibroblast growth factor receptor 3 (achondrop...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>hspa5</td>\n",
       "      <td>heat shock protein family A (Hsp70) member 5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>nr2f2</td>\n",
       "      <td>nuclear receptor subfamily 2 group F member 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>576</th>\n",
       "      <td>nr4a1</td>\n",
       "      <td>nuclear receptor subfamily 4 group A member 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    xeno_symbol                                       human_symbol\n",
       "7       csnk1a1                            casein kinase 1 alpha 1\n",
       "133       fgfr3  fibroblast growth factor receptor 3 (achondrop...\n",
       "223       hspa5       heat shock protein family A (Hsp70) member 5\n",
       "497       nr2f2      nuclear receptor subfamily 2 group F member 2\n",
       "576       nr4a1      nuclear receptor subfamily 4 group A member 1"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Inspect a few example mappings (Xenopus symbol -> human symbol)\n",
    "xeno_orth_sub[[\"xeno_symbol\", \"human_symbol\"]].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "22538c45-a637-41bc-a594-d964f99eb96c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "relationship\n",
       "1to1    101\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count how many distinct human symbols each Xenopus symbol maps to\n",
    "counts = xeno_orth_sub.groupby(\"xeno_symbol\")[\"human_symbol\"].nunique().rename(\"n_human\")\n",
    "\n",
    "# Add this count back into the table\n",
    "xeno_orth_sub = xeno_orth_sub.merge(counts, on=\"xeno_symbol\")\n",
    "\n",
    "# Classify relationships\n",
    "xeno_orth_sub[\"relationship\"] = np.where(xeno_orth_sub[\"n_human\"] == 1, \"1to1\", \"1toMany\")\n",
    "\n",
    "# How many genes are 1-to-1 vs 1-to-many?\n",
    "xeno_orth_sub[\"relationship\"].value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "01a316b4-4440-4629-8eea-794eb7a6b674",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(101,\n",
       " [('csnk1a1', 'casein kinase 1 alpha 1'),\n",
       "  ('fgfr3', 'fibroblast growth factor receptor 3 (achondroplasia'),\n",
       "  ('hspa5', 'heat shock protein family A (Hsp70) member 5'),\n",
       "  ('nr2f2', 'nuclear receptor subfamily 2 group F member 2'),\n",
       "  ('nr4a1', 'nuclear receptor subfamily 4 group A member 1'),\n",
       "  ('ruvbl2', 'RuvB like AAA ATPase 2'),\n",
       "  ('ptf1a', 'pancreas associated transcription factor 1a'),\n",
       "  ('lef1', 'lymphoid enhancer binding factor 1'),\n",
       "  ('sox5', 'SRY-box 5'),\n",
       "  ('smo', 'smoothened')])"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Keep only 1-to-1 orthologs\n",
    "orth_baseline = xeno_orth_sub[xeno_orth_sub[\"relationship\"] == \"1to1\"].copy()\n",
    "\n",
    "# In case of duplicates, keep one row per Xenopus symbol\n",
    "orth_baseline = orth_baseline.drop_duplicates(subset=[\"xeno_symbol\"])\n",
    "\n",
    "# Build dict: Xenopus symbol -> human symbol\n",
    "xeno_to_human_baseline = pd.Series(\n",
    "    orth_baseline[\"human_symbol\"].values,\n",
    "    index=orth_baseline[\"xeno_symbol\"].values\n",
    ").to_dict()\n",
    "\n",
    "len(xeno_to_human_baseline), list(xeno_to_human_baseline.items())[:10]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d49837d9-17ed-4a6a-819b-fde409b3922d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save as TSV\n",
    "orth_baseline.to_csv(\"../data/orthologs_1to1_xeno_human_101.tsv\",\n",
    "                     sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "79cc78ea-39a4-4969-88f6-c265b6b719bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "with open(\"../data/xeno_to_human_baseline.pkl\", \"wb\") as f:\n",
    "    pickle.dump(xeno_to_human_baseline, f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "521f682c-5ed1-49b4-98ca-d73998ca7af9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (xenopus_sc)",
   "language": "python",
   "name": "xenopus_sc"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.25"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
